src/Orchard.Web/Modules/Orchard.Widgets/Controllers/AdminController.cs

```diff
@@ -84,14 +86,27 @@ public class AdminController : Controller, IUpdateModel {
             string zonePreviewImagePath = string.Format("{0}/{1}/ThemeZonePreview.png", currentTheme.Location, currentTheme.Id);
             string zonePreviewImage = _virtualPathProvider.FileExists(zonePreviewImagePath) ? zonePreviewImagePath : null;
 
-            var widgets = _widgetsService.GetWidgets();
+            //Only when "ContentWidgets" layer is explicitly chosen the widgets are listed, not by default (for performance reasons)
+            IEnumerable<WidgetPart> widgets;
+            var IdLayerContentWidget = layers.Where(x => x.Name.ToLowerInvariant() == "contentwidgets").Select(x => x.Id).FirstOrDefault();
+            if (IdLayerContentWidget > 0) {
+                //Ok the layer exists
+                var IdsLayers = layers.Where(x => x.Name.ToLowerInvariant() != "contentwidgets").Select(x => x.Id).ToList().ToArray();
+                if (layerId != null && (int)layerId == IdLayerContentWidget) {
+                    //If ContentWidgets is chosen explicitly then IdsLayers is not filtered, note that the editors should be advised that this can take lot of time
+                    IdsLayers = layers.Select(x => x.Id).ToList().ToArray();
+                }
+                widgets = _widgetsService.GetWidgets(IdsLayers);
+            }
+            else {
+                widgets = _widgetsService.GetWidgets();
+            }
```

Comment:
Simplified logic with corrected variable naming.
```suggestion
            // Only listing the widgets of the "ContentWidgets" layer when it's explicitly chosen, not by default (for
            // performance reasons).
            IEnumerable<WidgetPart> widgets;
            var contentWidgetsLayer = layers.Where(x => x.Name.ToLowerInvariant() == "contentwidgets").FirstOrDefault();
            if (contentWidgetsLayer == null) {
                widgets = _widgetsService.GetWidgets();
            }
            else { // The "ContentWidgets" layer exists.
                var layerIds = layers.Select(x => x.Id).ToList();
                // There's no layer selected or it's not "ContentWidgets", so excluding it from the widget listing.
                if (!layerId.Equals(contentWidgetsLayer.Id)) {
                    layerIds.Remove(contentWidgetsLayer.Id);
                }

                widgets = _widgetsService.GetWidgets(layerIds.ToArray());
            }
```